Implement S3 Gateway to third party cloud storage providers.

krishnasrinivas opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Currently supported backend is Azure Blob Storage.

```
export MINIO_ACCESS_KEY=azureaccountname
export MINIO_SECRET_KEY=azureaccountkey
minio gateway azure
```

<!--- Describe your changes in detail -->

## Motivation and Context
Minio gateway adds Amazon S3 compatibility to third party cloud storage providers.

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
Manually and by rest of the @minio/core 
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

krishnasrinivas committed:
azure: Azure blob storage backend for minio.

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,288 @@
+	for _, container := range resp.Containers {
+		t, e := time.Parse(time.RFC1123, container.Properties.LastModified)
+		if e != nil {
+			return nil, traceError(e)
```
Should we exit on error or continue to the next iteration?

krishnasrinivas replied:
return/continue anything is fine.
but parsing should never fail here, if it does then something very basic has failed and better to return an error else the error gets ignored.

krishnasrinivas committed:
Vendorize github.com/Azure/azure-sdk-for-go/storage

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,339 @@
+
+func newAzureLayer(account, key string) (ObjectLayer, error) {
+	useHTTPS := true
+	apiVersion := "2016-05-31"
```
Make this a const.

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,339 @@
+type AzureObjects struct {
+	client storage.BlobStorageClient // Azure sdk client
+}
+
```
Comment needed for `newAzureLayer`.

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,339 @@
+	return &AzureObjects{client}, nil
+}
+
+// Shutdown - Not relavant to Azure.
```
s/relavant/relevent.

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,339 @@
+	return nil
+}
+
+// StorageInfo - Not relavant to Azure.
```
s/relavant/relevant

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,339 @@
+	"github.com/Azure/azure-sdk-for-go/storage"
+)
+
+// AzureObjects - Azure Object layer
```
A comment explaining which of the azure storage services this layer is mapping to would help. E.g, we are using Azure blob storage here.

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,339 @@
+		}
+		return traceError(err)
+	}
+	io.Copy(writer, rc)
```
We need to handle error returned by `io.Copy`.

krishnasrinivas replied:
once we start writing to the client we won't be able to handle the error, i.e we would have already sent http status 200, at best we can log an error.

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,339 @@
+func (a AzureObjects) GetObjectInfo(bucket, object string) (objInfo ObjectInfo, err error) {
+	prop, err := a.client.GetBlobProperties(bucket, object)
+	if err != nil {
+		if e, ok := err.(storage.AzureStorageServiceError); ok {
```
Getting `AzureStorageServiceError` could be moved into a function.

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,339 @@
+		return result, nil
+	}
+	if len(resp.Parts) > 0 {
+		result.Uploads = []uploadMetadata{{prefix, prefix, time.Now(), ""}}
```
Shouldn't we return `time.Now().UTC()` or is it in Azure spec to return time based on locale?

krishnasrinivas replied:
time.Now().UTC() is better
note that it is more of S3 spec, azure does not store the modified time, hence we are just filling in a time. an S3 client that depends on this time will not work as expected

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,339 @@
+
+// NewMultipartUpload - Use Azure equivalent CreateBlockBlob.
+func (a AzureObjects) NewMultipartUpload(bucket, object string, metadata map[string]string) (uploadID string, err error) {
+	uploadID = object
```
Could you add a comment explaining why object name is being used as uploadID?

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,339 @@
+	}
+	info.PartNumber = partID
+	info.ETag = md5Hex
+	info.LastModified = time.Now()
```
Is the time based on minio server locale or  should it be UTC?

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,339 @@
+		}
+		result.Parts = append(result.Parts, PartInfo{
+			partID,
+			time.Now(),
```
Same comment as above.

krisis commented on cmd/credential.go:
```diff
@@ -30,7 +30,7 @@ const (
 	accessKeyMaxLen = 20
 	secretKeyMinLen = 8
-	secretKeyMaxLen = 40
+	secretKeyMaxLen = 100
```
Shouldn't this be changed only for Azure and left as is for S3?

harshavardhana replied:
:+1:

krishnasrinivas replied:
problem with having one value for S3 and one for azure is we will have to maintain the logic/code for it.

i.e we can declare global variable

var secretKeyMaxLen

this would have to be initialized to 80 for S3 and 100 for azure and also maintain different test cases for each. Simpler solution would be to keep it as one value.

We also have to make this change to mc, in mc there is no way to know if it is azure in the backend.

harshavardhana replied:
We can keep validation techniques more generalized so that its easier.

vadmeste commented on cmd/azure.go:
```diff
@@ -0,0 +1,339 @@
+		return objInfo, traceError(err)
+	}
+	objInfo, err = a.GetObjectInfo(bucket, object)
+	if err != nil {
```
This is an empty if block {}

krishnasrinivas committed:
Fix for review comments.

harshavardhana commented on cmd/azure.go:
```diff
@@ -12,28 +12,40 @@ import (
 )
 
-// AzureObjects - Azure Object layer
+const azureAPIVersion = "2016-05-31"
```
Append with 'global'

harshavardhana commented on cmd/azure.go:
```diff
@@ -12,28 +12,40 @@ import (
 
+// Convert azure errors to minio object layer errors.
+func azureToObjectError(err error, bucket, object string) error {
+	if e, ok := err.(storage.AzureStorageServiceError); ok {
```
Perhaps bring a switch case which can be expanded for other errors as well.

krishnasrinivas committed:
Set secretKeyMaxLen = 100 for Azure backend.

krishnasrinivas committed:
Refactor azureToObjectError to use switch statement.

krisis commented on cmd/azure.go:
```diff
@@ -147,12 +158,7 @@ func (a AzureObjects) GetObject(bucket, object string, startOffset int64, length
-			}
-		}
-		return objInfo, traceError(err)
+		return objInfo, traceError(azureToObjectError(err, bucket, object))
```
same as above comment.

krisis commented on cmd/azure.go:
```diff
@@ -131,12 +147,7 @@ func (a AzureObjects) GetObject(bucket, object string, startOffset int64, length
-			}
-		}
-		return traceError(err)
+		return traceError(azureToObjectError(err, bucket, object))
```
shouldn't it be `azureToObjectError(traceError(err), bucket, object)`?

krishnasrinivas replied:
it will not matter because traceError captures the line number in either case.

if it's about consistency, we can follow `azureToObjectError(traceError(err), bucket, object)?`

krisis replied:
Yes, I made that suggestion to be consistent with our practice in rest of the code.

krisis commented on cmd/azure.go:
```diff
@@ -216,13 +212,14 @@ func (a AzureObjects) ListMultipartUploads(bucket, prefix, keyMarker, uploadIDMa
 
 // NewMultipartUpload - Use Azure equivalent CreateBlockBlob.
 func (a AzureObjects) NewMultipartUpload(bucket, object string, metadata map[string]string) (uploadID string, err error) {
+	// Azure does not support multiple upload ids for an object. Hence we use object name itself as the uploadID.
```
So, what happens when we try to upload multipart objects concurrently. Does NewMultipartUpload fail with upload already exists if an earlier request successfully created one?

krishnasrinivas replied:
since we call azure directly, depends on how azure behaves. we need not worry about it because we inherit azure's limitations.

krisis commented on cmd/credential.go:
```diff
@@ -30,12 +30,13 @@ const (
 	alphaNumericTableLen = byte(len(alphaNumericTable))
 )
 
+var secretKeyMaxLen = 40
```
- we should use different constants for Azure and AWS S3 maximum secret key/access key length instead of literals.
- as per our convention, we should call the above variable globalSecretKeyMaxLen.

krishnasrinivas replied:
> we should use different constants for Azure and AWS S3 maximum secret key/access key length instead of literals.

keeping different constants creates problem as the signature verification code is the same. we'll solve this differently as there is a refactor planned, not sure of the complete picture yet.

> as per our convention, we should call the above variable globalSecretKeyMaxLen.

other consts in credential.go does not have global prefix hence kept it this way. there is inconsistency here.

krisis replied:
>keeping different constants creates problem as the signature verification code is the same. we'll solve this differently as there is a refactor planned, not sure of the complete picture yet.

I think I wasn't clear. I am suggesting we use different constants like,
```go
const  (
  globalAzureMaxSecretKeyLen = 100
  globalS3MaxSecretKeyLen = 40
)
```
instead of using the integer literals `40` and `100` directly. 

You can still use `secretMaxKeyLen` with or without global prefixed to it, which will be used by the signature code.

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,333 @@
+		MaxResults: uint(maxKeys),
+	})
+	if err != nil {
+		return result, traceError(err)
```
Shouldn't this be, `return result, azureToObjectError(traceError(err), bucket, prefix)`.

krishnasrinivas replied:
yes, we can do that. i started adding `azureToObjectError(traceError(err), bucket, prefix)` only on need basis. but i will make this change.

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,333 @@
+	if err != nil {
+		return traceError(azureToObjectError(err, bucket, object))
+	}
+	io.Copy(writer, rc)
```
We should handle error returned by `io.Copy` and return it to the user.

krishnasrinivas replied:
true, if no data is written yet then s3 layer returns non 200 status, will do that change.

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,333 @@
+	}
+	t, err := time.Parse(time.RFC1123, prop.LastModified)
+	if err != nil {
+		return objInfo, traceError(err)
```
Shouldn't this be, `return ObjectInfo{}, azureToObjectError(traceError(err), bucket, object)`?

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,333 @@
+	result.Delimiter = delimiter
+	resp, err := a.ListObjectParts(bucket, prefix, prefix, 1, 1000)
+	if err != nil {
+		return result, nil
```
why are we returning `nil` on finding an error?

krishnasrinivas replied:
In case ListObjectParts returns error, it would mean that no such incomplete upload exists on azure storage - in which case we return nil. This plays a role for mc and our client libs. i.e client does ListMultipartUploads to figure the previous uploadID to continue uploading from where it left off. If we return error the upload never starts and always returns error.

krisis replied:
Could you add this in a comment?

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,333 @@
+
+// NewMultipartUpload - Use Azure equivalent CreateBlockBlob.
+func (a AzureObjects) NewMultipartUpload(bucket, object string, metadata map[string]string) (uploadID string, err error) {
+	// Azure does not support multiple upload ids for an object. Hence we use object name itself as the uploadID.
```
Azure blob storage (ASB) doesn't return a unique multipart upload ID like AWS  S3. It places the burden of creating unique blockIDs for the blocks uploaded to a given object on the user. So, we could update the comment to say that Azure doesn't return a unique upload ID and we use object name in place of it. ASB allows multiple uploads to co-exist as long as the user keeps the blocks uploaded (in block blobs) unique amongst concurrent upload attempts. Each concurrent client, keeps its own blockID list which it can commit.

krishnasrinivas replied:
well explained, i will update the comment with the same content.

though we can use "azure" instead of ASB as in our context azure means azure-blob-storage and more easily readable.

krishnasrinivas committed:
Fix review comments.

krishnasrinivas committed:
Fixes needed after rebase.

donatello commented on cmd/azure.go:
```diff
@@ -0,0 +1,363 @@
+// GetBucketInfo - Get bucket metadata.
+func (a AzureObjects) GetBucketInfo(bucket string) (BucketInfo, error) {
+	// Azure does not have an equivalent call, hence use ListContainers.
+	resp, err := a.client.ListContainers(storage.ListContainersParameters{
```
Azure does provide an equivalent call for `GetBucketInfo` - it is called get-container-properties - https://docs.microsoft.com/en-us/rest/api/storageservices/fileservices/get-container-properties

This implementation can be significantly simplified using this API rather than listing containers and searching.

krishnasrinivas replied:
yes, but it is not available in azure-sdk-for-go

donatello commented on cmd/azure.go:
```diff
@@ -0,0 +1,363 @@
+	byteRange := fmt.Sprintf("%d-%d", startOffset, startOffset+length-1)
+	if length == -1 {
+		byteRange = fmt.Sprintf("%d-", startOffset)
+	}
```
The above 4 lines is better done by assigning `byteRange` only once. Here in the case when length = -1, it will perform two Sprintf calls.

krishnasrinivas replied:
will fix that

donatello commented on cmd/azure.go:
```diff
@@ -0,0 +1,363 @@
+}
+
+// GetObject - Use Azure equivalent GetBlobRange.
+func (a AzureObjects) GetObject(bucket, object string, startOffset int64, length int64, writer io.Writer) (err error) {
```
Note of caution - according to https://docs.microsoft.com/en-us/rest/api/storageservices/fileservices/specifying-the-range-header-for-blob-service-operations - it is not apparent if the server supports a `Range` header larger than 4GB (and this what the underlying azure go library seems to be passing to the service):

> Several Blob service GET operations support the use of the standard HTTP Range header. Many HTTP clients, including the .NET client library, limit the size of the Range header to a 32-bit integer, and thus its value is limited to a maximum of 4 GB. Since both block blobs and page blobs can be larger than 4 GB in size, the Blob service accepts a custom range header x-ms-range for any operation that takes an HTTP Range header.

krishnasrinivas replied:
this is fine, we inherit the limitations of azure and its SDK. we'll be a minimal layer on top of azure unless it becomes a concern later that needs to be fixed

krishnasrinivas replied:
i see this in the sdk:
```
        headers := b.client.getStandardHeaders()                                                                                                                                                                                                                                                                                                                                            
        if bytesRange != "" {                                                                                                                                                                                                                                                                                                                                                               
                headers["Range"] = fmt.Sprintf("bytes=%s", bytesRange)                                                                                                                                                                                                                                                                                                                      
        }             
```
so we are fine.

donatello replied:
@krishnasrinivas That is just saying that the library is using `Range` header, for which their docs is not clear about supporting a range larger than 4GB (endoffset - startoffset > 4GB). They support the header `x-ms-range` for which the server does support sending a header larger than 4GB (see the API doc link I put above).

So I don't understand why you say its fine.

harshavardhana replied:
Since Azure SDK for Go doesn't support it we don't either, we are not writing anything else on top of it.  If Azure SDK ends up supporting this in future then we would inherit that too and we don't have to do anything.  If Azure library is short handed and doesn't even implement the features they have its not our problem to address them. As and when the library catches up to their own features we will transparently support it. 

It is similar to how `minio-go` supports GCS (Google Cloud Storage) we don't even attempt to do multipart since GCS implements an S3 incompatible implementation.

donatello replied:
> As and when the library catches up to their own features we will transparently support it.

Sure, I agree with this view, same as what @krishnasrinivas said in the first response. Just didnt understand his second response.

donatello commented on cmd/azure.go:
```diff
@@ -0,0 +1,363 @@
+}
+
+// PutObject - Use Azure equivalent CreateBlockBlobFromReader.
+func (a AzureObjects) PutObject(bucket, object string, size int64, data io.Reader, metadata map[string]string, sha256sum string) (objInfo ObjectInfo, err error) {
```
Please wrap/break the line for readability - I understand go does not _require_ you to do so, but it is easier to read. Right now the line is falling off the margins on github. Left like this it is easy to miss bugs like unused arguments, shadowing of variables (like in this serious bug - https://github.com/minio/minio/pull/3838).

krishnasrinivas replied:
The line wrap issue can be taken care of in a separate fix because:
```
krishna@escape:~/gocode/src/github.com/minio/minio/cmd$ cat * | awk 'length($0) > 100' | wc -l
1996
krishna@escape:~/gocode/src/github.com/minio/minio/cmd$ 

```

krisis commented on cmd/azure.go:
```diff
@@ -0,0 +1,363 @@
+// Encode partID+md5Hex to a blockID.
+func azureGetBlockID(partID int, md5Hex string) string {
+	return base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%.5d.%s", partID, md5Hex)))
+}
```
This is incorrect - ```fmt.Sprintf("%.5d.%s", partID, md5Hex))```
the format-specifier for partID should be "%06d", since there can be upto 100,000 uncommitted blocks and it the blockID as a whole is required to be equal length, so we need to pad the part number with leading zeros.

krishnasrinivas replied:
maximum part number that S3 clients will send is 10000, s3 layer does not allow putobjectpart with partnumber > 10000

donatello replied:
> maximum part number that S3 clients will send is 10000, s3 layer does not allow putobjectpart with partnumber > 10000

Minor point but `"%.5d"` is slight abuse of the format string. After the period, the five becomes precision (meant to be used for floats) though we mean width here. Though this appears to work, it is not implied by the spec https://golang.org/pkg/fmt/#hdr-Printing. Instead "%05d" is suggested.

donatello commented on cmd/azure.go:
```diff
@@ -0,0 +1,363 @@
+}
+
+// AbortMultipartUpload - Not Implemented.
+func (a AzureObjects) AbortMultipartUpload(bucket, object, uploadID string) error {
```
Uncommitted blocks in a blob are automatically deleted after a week. However, for reference, in case immediate removal of uncommitted blocks is desired (as they are charged for a week), one way I could find was using copy-object with source name = destination name:

> You can copy a source blob to a destination blob with the same name, effectively replacing the destination blob. Such a copy operation removes any uncommitted blocks and overwrites the blob's metadata.

from [here](https://docs.microsoft.com/en-us/rest/api/storageservices/fileservices/copy-blob#remarks).

krishnasrinivas replied:
Does their doc say that they charge for uncommitted blocks?



> You can copy a source blob to a destination blob with the same name, effectively replacing the destination blob. Such a copy operation removes any uncommitted blocks and overwrites the blob's metadata.

It is a risky to do it as it will affect the namespace (for example any existing object by that name getting deleted)

donatello commented on cmd/azure.go:
```diff
@@ -0,0 +1,363 @@
+func azureToObjectError(err error, params ...string) error {
+	bucket := ""
+	object := ""
+	if len(params) == 1 {
```
`if len(params) >= 1 {`

krishnasrinivas replied:
will fix that.

krishnasrinivas committed:
Add license header.

harshavardhana commented on cmd/azure.go:
```diff
@@ -0,0 +1,363 @@
+package cmd
```
Add license header..

krisis replied:
License header nit: s/2016/2017

harshavardhana commented on cmd/azure.go:
```diff
@@ -0,0 +1,363 @@
+	result.MaxUploads = maxUploads
+	result.Prefix = prefix
+	result.Delimiter = delimiter
+	resp, err := a.ListObjectParts(bucket, prefix, prefix, 1, 1000)
```
Shouldn't 1000 be 'maxUploads' here?

krishnasrinivas replied:
note the comment:
```
// ListMultipartUploads - Incomplete implementation, for now just return the prefix if it is an incomplete upload.
```

ListMultipartUploads() isn't fully implemented yet - good enough for the demo. Need to explore if at all it can be implemented fully because azure does not expose a suitable API for list-incomplete-uploads.

harshavardhana replied:
> ListMultipartUploads() isn't fully implemented yet - good enough for the demo. Need to explore if at all it can be implemented fully because azure does not expose a suitable API for list-incomplete-uploads.

Ah i see..

donatello commented on cmd/azure.go:
```diff
@@ -0,0 +1,385 @@
+	result.Bucket = bucket
+	result.Object = object
+	result.UploadID = uploadID
+	result.MaxParts = maxParts
```
Don't understand why `maxParts` is intentionally ignored here. Or is it a bug?

Azure returns uncommitted blobs in alphabetical blockid order. It can be taken advantage of to do ListObjectParts (without any concurrency guarantees, just like Azure itself) respecting `maxParts`. [Just send the offset of the next block in the list as part-number-marker in the first response, and in the next pagination request, it will be sent back by the client, we can use it to return the next `maxParts` blocks].

krishnasrinivas replied:
good catch, it's a bug. during development i thought about fixing it later which i never did.

donatello replied:
`NextPartNumberMarker` needs to be set in the response.

krishnasrinivas committed:
Fix ListObjectParts to support the case where more than 1000 parts are uploaded.

krishnasrinivas committed:
Implement bucket policies for for gateway mode.

krishnasrinivas committed:
Fix Travis warnings.

harshavardhana commented on cmd/azure.go:
```diff
@@ -0,0 +1,515 @@
+/*
+ * Minio Cloud Storage, (C) 2016 Minio, Inc.
```
License header should be 2017.

harshavardhana commented on cmd/azure.go:
```diff
@@ -0,0 +1,515 @@
+	objInfo.ModTime = t
+	objInfo.Name = object
+	objInfo.Size = prop.ContentLength
+	return
```
Providing an explicit return is more readable.

harshavardhana commented on cmd/azure.go:
```diff
@@ -0,0 +1,515 @@
+	return a.GetObjectInfo(bucket, object)
+}
+
+// HealBucket - Not relevant.
```
Move this to un-implemented `azure-unsupported.go` - this can be purged when you bring in the new interface.

krishnasrinivas replied:
Actually, I have made the interface changes in another branch and purged these functions.

harshavardhana replied:
No meaning even in this PR we can move these to a separate file so that they are not written in-between and makes it easy to purge them.  For the next rebase.

harshavardhana reviewed (changes requested):
Some minor changes..

krishnasrinivas committed:
Review fixes.

harshavardhana commented on cmd/gateway-main.go:
```diff
@@ -0,0 +1,101 @@
+	Subcommands: []cli.Command{
+		{
+			Name:  "azure",
+			Usage: "Azure cloud",
```
We need a helpTemplate here just like serverMain() to provide examples for how to run azure.. Can you add these examples? 

Also can you create a simple markdown documenting the gateway changes perhaps ideally in docs/gateway directory top-level README.md would do.

krishnasrinivas replied:
I will push the markdown doc.

krishnasrinivas committed:
Move unsupported Heal related functions to azure-unsupported.go

krishnasrinivas committed:
Provide better help message on the command line.

harshavardhana commented on cmd/gateway-main.go:
```diff
@@ -21,33 +21,51 @@ import (
-func gatewayMain(ctx *cli.Context) {
-	cli.ShowSubcommandHelp(ctx)
-}
+// // Handler for 'minio gateway' - show help.
```
You have to remove this not needed anymore.

harshavardhana committed:
Formatting changes and some texts.

krishnasrinivas committed:
Add markdown doc for running minio in gateway mode.

harshavardhana commented on docs/gateway.md:
```diff
@@ -0,0 +1,13 @@
+Minio can be run in the gateway mode where it will be a proxy to backend cloud storage providers. Currently Minio supports Azure backend.
+
+## Running in Gateway mode
+
```
Mention where they can get the gateway keys from @krishnasrinivas  - perhaps a azure link.

harshavardhana commented on docs/gateway.md:
````diff
@@ -0,0 +1,13 @@
+$ export MINIO_SECRET_KEY=azureaccountkey
+$ minio gateway azure
+```
+
````
You can detail out `mc config host add` so that users can directly use it.

krishnasrinivas committed:
Remove stale commented code.

harshavardhana committed:
Add more changes with doc updates.

Also re-factor to not use `gateway` config.json,
instead instantiate config separately.

Add support for creating certs config dir for gateway.

krishnasrinivas committed:
Rebase fixes.

krishnasrinivas committed:
Disable bucket notification handlers.

krishnasrinivas committed:
Override DeleteBucketHandler for gateway.

harshavardhana committed:
Enable console logging.

krishnasrinivas committed:
Fix azureToObjectError()

krishnasrinivas committed:
Remove few checks that were added by mistake.

krishnasrinivas committed:
Add anonymous list-objects support.

krishnasrinivas committed:
Rename variables

krishnasrinivas committed:
Implement GetBucketLocationHandler for gateway handler to handle anonymous request.

krishnasrinivas committed:
Handle Range header correctly.

harshavardhana committed:
docs: update gateway description and instructions.

harshavardhana committed:
Bunch of fixes after running karthic's mint tests - still partial.

krishnasrinivas committed:
Make azureToObjectError readable.

krishnasrinivas committed:
Fix GET object of empty files.

harshavardhana committed:
Handle anonymous getObject httpStatus properly.

krishnasrinivas committed:
Mint fixes.

krishnasrinivas committed:
For multi-object-delete delete objects serially.

Azure resets connection if there are too many active connections.

krishnasrinivas committed:
Support for abort-multipart-upload and metadata for multipart uploads.

harshavardhana commented on cmd/azure.go:
```diff
@@ -103,7 +130,8 @@ func newAzureLayer(account, key string) (GatewayLayer, error) {
 	}
 	client := c.GetBlobService()
-	return &AzureObjects{client}, nil
+	meta := azureMultipart{meta: make(map[string]map[string]string)}
```
You don't need a new variable. Two meta in same line is confusing. You can move this directly to the next line..

krishnasrinivas replied:
ok

harshavardhana commented on cmd/azure.go:
```diff
@@ -38,9 +39,35 @@ import (
 
+// To store metadata during NewMultipartUpload which will be used after
+// CompleteMultipartUpload to call SetBlobMetadata.
+type azureMultipart struct {
```
You can call this azureMultipartInfo....

krishnasrinivas replied:
will call it `azureMultipartMetaInfo` but it looks long for a type, calling it `azureMultipartInfo` will be incorrect as it does not hold info related to multipart but rather meta of the multipart.

harshavardhana commented on cmd/azure.go:
```diff
@@ -38,9 +39,35 @@ import (
+	mu   sync.Mutex
+}
+
+func (a *azureMultipart) get(key string) map[string]string {
```
Need a comment for each line..

krishnasrinivas replied:
will add it

harshavardhana commented on cmd/azure.go:
```diff
@@ -38,9 +39,35 @@ import (
+// CompleteMultipartUpload to call SetBlobMetadata.
+type azureMultipart struct {
+	meta map[string]map[string]string
+	mu   sync.Mutex
```
Why not use NSlock in respective calls ? Avoids having more locks here which are not needed ?.

krishnasrinivas replied:
since this lock is for protecting the map it would not be right to force fit nslock here.
have a look at the get API:
```
// To store metadata during NewMultipartUpload which will be used after                                                                     
// CompleteMultipartUpload to call SetBlobMetadata.                                                                                         
type azureMultipartMetaInfo struct {                                                                                                        
        meta map[string]map[string]string                                                                                                   
        sync.Mutex                                                                                                                          
}                                                                                                                                           
                                                                                                                                            
func (a *azureMultipartMetaInfo) get(key string) map[string]string {                                                                        
        a.Lock()                                                                                                                            
        defer a.Unlock()                                                                                                                    
        return a.meta[key]                                                                                                                  
}
```
this will look like:
```
// To store metadata during NewMultipartUpload which will be used after                                                                     
// CompleteMultipartUpload to call SetBlobMetadata.                                                                                         
type azureMultipartMetaInfo struct {                                                                                                        
        meta map[string]map[string]string                                                                                                   
}                                                                                                                                           
                                                                                                                                            
func (a *azureMultipartMetaInfo) get(key, bucket, object string) map[string]string {                                                                        
        objectLock := globalNSMutex.NewNSLock(bucket, object)                                                                               
        objectLock.RLock()                                                                                                                  
        defer objectLock.RUnlock()
        return a.meta[key]                                                                                                                  
}
```

harshavardhana commented on cmd/azure.go:
```diff
@@ -29,6 +29,7 @@ import (
 	"net/url"
 	"strconv"
```
One more thing to notice here is we are not persisting metadata and can get lost upon gateway restart, so i am not sure why we even need to attempt saving metadata in memory.. ? because if the idea is to address the situation of preserving content-type we should indeed behave like how an S3 would behave.. i.e even upon a future completion of the upload it should save the content-type. 

Otherwise this change makes it inconsistent in nature where users need to be sure that it won't save content-type if server just happens to have been restarted..

krishnasrinivas replied:
yes, but there is no other choice, ab suggested this solution and say that multipart is not supported if server is restarted.
what can be done is store the metadata on a tmp `uuid` object and use it during complete multipart, will check with ab if that can be done. if you can think of any other way let me know.

krishnasrinivas replied:
ab suggested that `mc admin restart` can persist metadata on disk and load it again - after the release.
also have a expiry go-routine to purge old map entries - after the release.

harshavardhana replied:
okay..

harshavardhana replied:
BTW `admin` APIs are not registered with gateway and also it doesn't make sense to have admin APIs for gateway apart from `admin restart` and `status` to get uptime. 

All calls require underlying real object layer because .. 

- listing locks, clearing locks
- set/getconfig, changing password.
- healing,

krishnasrinivas committed:
Fix bug in the previous commit to use SetBlobProperties.

SetBolbMetadata does not work for Content-Type.

krishnasrinivas committed:
Rename variables and add comments.

krishnasrinivas committed:
Allow CompleteMultipartUpload even if meta map entry is not found.

krishnasrinivas committed:
Revert "Allow CompleteMultipartUpload even if meta map entry is not found."

This reverts commit c00a592cf92f74c955f73dd426b81d0c45da2f7b.

krishnasrinivas committed:
If entry not found in meta-map then return error for PutObjectPart and return empty list for ListObjectParts

krishnasrinivas committed:
Set NextPartNumberMarker in ListObjectParts response.

krishnasrinivas committed:
Fix test cases for azureToObjectError

harshavardhana committed:
Add more test cases, some comments.

Move anonymous calls into azure-anonymous.go

krishnasrinivas committed:
Add checkUpdate support for gateway.

harshavardhana commented on cmd/gateway-main.go:
```diff
@@ -207,6 +207,7 @@ func gatewayMain(ctx *cli.Context) {
 
 	// Prints the formatted startup message once object layer is initialized.
 	if !quietFlag {
+		checkUpdate(globalMinioModeGatewayAzure)
```
perhaps you can add this based on the backendType.. 

```
if backendType == azureBackend {
```

krishnasrinivas committed:
Minor refactor for gateway checkUpdate()

krishnasrinivas committed:
Fix compilation error.

krishnasrinivas committed:
ListMultipartUploads returns empty list if there is no entry in the meta-map.

krishnasrinivas committed:
Remove leading and trailing '"' from ETag.

harshavardhana commented on cmd/azure.go:
```diff
@@ -277,7 +277,7 @@ func (a AzureObjects) GetObjectInfo(bucket, object string) (objInfo ObjectInfo,
 		Bucket:      bucket,
 		UserDefined: make(map[string]string),
-		MD5Sum:      prop.Etag,
+		MD5Sum:      strings.Trim(prop.Etag, "\""),
```
We can use `canonicalizeETag` here..

krishnasrinivas committed:
Use canonicalizeETag()

krishnasrinivas committed:
Add test cases.

krishnasrinivas committed:
Fix test cases.

krishnasrinivas committed:
Fix test cases.

krishnasrinivas committed:
Rebase fixes.

krishnasrinivas committed:
Add gateway-azure-limitations.md

harshavardhana commented on docs/gateway/gateway-azure-limitations.md:
```diff
@@ -0,0 +1,20 @@
+- Non-empty buckets get removed on a DeleteBucket() call.
+
+Other limitations:
+- Current implementation of ListMultipartUploads is incomplete. Right now it returns if the "prefix"
```
This sentence needs to be fixed 
```
Current implementation of ListMultipartUploads is incomplete. Right now it returns if the object name "prefix" has any uploaded parts, otherwise returns empty. 
```

```
Bucket notifications are not supported.
```

krishnasrinivas committed:
Fix review comments.
